flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -446,9 +446,13 @@ private CleanupJobState handleJobManagerRunnerResult(
             JobManagerRunnerResult jobManagerRunnerResult, ExecutionType executionType) {
         if (jobManagerRunnerResult.isInitializationFailure()) {
             if (executionType == ExecutionType.RECOVERY) {
-                return jobManagerRunnerFailed(
-                        jobManagerRunnerResult.getExecutionGraphInfo().getJobId(),
-                        jobManagerRunnerResult.getInitializationFailure());
+                if (jobManagerRunnerResult.isJobFinishing()) {
+                    return CleanupJobState.GLOBAL;
```

Comment:
I think we should call `jobReachedTerminalState` so that we create an `ArchivedExecutionGraph`.

Reply:
I have implemented this by complete the jobManagerRunner resultFuture with 

```java
resultFuture.complete(
        JobManagerRunnerResult.forSuccess(
                new ExecutionGraphInfo(
                        jobMasterServiceProcessFactory.createArchivedExecutionGraph(
                                JobStatus.FAILED,
                                new JobAlreadyDoneException(getJobID())))));
```

But, I'm not sure is it right to set state to `JobStatus.FAILED`. We do not know the status of job at the time of recovery of an already done job. I think we can do these things like :

- Add a special `JobStatus.UNKNOWN` to handle this case
- Or when the FLIP for [FLINK-11813](https://issues.apache.org/jira/browse/FLINK-11813) have been introduced , we can extract the job result directly from `JobResultStore`.

Looking forward to your opinion.

Reply:
I would refrain from adding a new `JobStatus.UNKNOWN` value because `JobStatus` is public evolving and exposed at many places. I agree that `FAILED` is not optimal but we should state in the exception message that we could not properly deduce the final state. Moreover, I hope that this can be fixed with FLINK-11813 properly.